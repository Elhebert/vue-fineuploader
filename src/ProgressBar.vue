<template>
  <component
    :is="progressBar"
    :id="id"
    :uploader="uploader"
  >
    <div
      slot-scope="{ uploading, progress }"
      :hidden="!uploading && hidesWhenNotUploading"
      :aria-valuenow="progress"
      :style="{ 'width': `${progress}%` }"
      aria-valuemax="100"
      aria-valuemin="0"
      role="progressbar" />
  </component>
</template>

<script>
import RenderlessFileProgressBar from './renderless/FileProgressBar'
import RenderlessTotalProgressBar from './renderless/TotalProgressBar'

export default {
  components: {
    RenderlessFileProgressBar,
    RenderlessTotalProgressBar,
  },
  props: {
    id: {
      type: Number,
      default: null,
    },
    hidesWhenNotUploading: {
      type: Boolean,
      default: true,
    },
    uploader: {
      type: Object,
      required: true,
    },
  },

  computed: {
    progressBar() {
      return this.id === null
        ? 'renderless-total-progress-bar'
        : 'renderless-file-progress-bar'
    },
  },
}
</script>
