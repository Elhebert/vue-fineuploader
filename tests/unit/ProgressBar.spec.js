import { mount } from '@vue/test-utils'
import ProgressBar from '@/ProgressBar.vue'
import FineUploader from 'fine-uploader-wrappers'

describe('ProgressBar.vue', () => {
  it('hides the progress bar by default', () => {
    const uploader = new FineUploader({ options: {} })
    let wrapper = mount(ProgressBar, { propsData: { id: 0, uploader } })

    expect(wrapper.attributes().hidden).toBeDefined()
    expect(wrapper).toMatchSnapshot()

    wrapper = mount(ProgressBar, { propsData: { uploader } })

    expect(wrapper.attributes().hidden).toBeDefined()
    expect(wrapper).toMatchSnapshot()

  })

  it('shows the file progress bar before uploading start if specified', () => {
    const uploader = new FineUploader({ options: {} })
    let wrapper = mount(ProgressBar, {
      propsData: { id: 0, uploader, hidesWhenNotUploading: false },
    })

    expect(wrapper.attributes().hidden).toBeUndefined()
    expect(wrapper).toMatchSnapshot()
  })

  it('shows the total progress bar before uploading start if specified', () => {
    const uploader = new FineUploader({ options: {} })
    let wrapper = mount(ProgressBar, {
      propsData: { uploader, hidesWhenNotUploading: false },
    })

    expect(wrapper.attributes().hidden).toBeUndefined()
    expect(wrapper).toMatchSnapshot()
  })

  it('show the file progress if an ID is supplied', () => {
    const uploader = new FineUploader({ options: {} })
    const wrapper = mount(ProgressBar, { propsData: { id: 3, uploader } })
    const renderless = wrapper.vm.$children[0]

    renderless.onProgress(3, 'foo.jpeg', 100, 1000)

    expect(wrapper).toMatchSnapshot()
    expect(wrapper.attributes().style).toBe('width: 10%;')
  })

  it('show the total progress if no ID is supplied', () => {
    const uploader = new FineUploader({ options: {} })
    const wrapper = mount(ProgressBar, { propsData: { uploader } })
    const renderless = wrapper.vm.$children[0]

    renderless.onProgress(100, 1000)

    expect(wrapper).toMatchSnapshot()
    expect(wrapper.attributes().style).toBe('width: 10%;')
  })

  it('hides total progress bar after all uploads are complete', () => {
    const uploader = new FineUploader({ options: {} })
    const wrapper = mount(ProgressBar, { propsData: { uploader } })
    const renderless = wrapper.vm.$children[0]

    // uploading
    uploader.methods['getInProgress'] = jest.fn(() => 1)
    renderless.onStatusChange(
      0,
      uploader.qq.status.QUEUED,
      uploader.qq.status.UPLOADING,
    )

    expect(wrapper).toMatchSnapshot()
    expect(wrapper.attributes().hidden).toBeUndefined()

    // done uploading
    uploader.methods['getInProgress'] = jest.fn(() => 0)
    renderless.onStatusChange(
      0,
      uploader.qq.status.UPLOADING,
      uploader.qq.status.UPLOAD_SUCCESSFUL,
    )

    expect(wrapper).toMatchSnapshot()
    expect(wrapper.attributes().hidden).toBeDefined()
  })

  it('hides file progress bar after all upload is complete', () => {
    const uploader = new FineUploader({ options: {} })
    const wrapper = mount(ProgressBar, { propsData: { id: 0, uploader } })
    const renderless = wrapper.vm.$children[0]

    // uploading
    renderless.onStatusChange(0, uploader.qq.status.QUEUED, uploader.qq.status.UPLOADING)

    expect(wrapper).toMatchSnapshot()
    expect(wrapper.attributes().hidden).toBeUndefined()

    // done uploading
    renderless.onStatusChange(0, uploader.qq.status.UPLOADING, uploader.qq.status.UPLOAD_SUCCESSFUL)

    expect(wrapper).toMatchSnapshot()
    expect(wrapper.attributes().hidden).toBeDefined()
  })

  it('shows total progress bar after all uploads are complete if requested', () => {
    const uploader = new FineUploader({ options: {} })
    const wrapper = mount(ProgressBar, {
      propsData: { uploader, hidesWhenNotUploading: false },
    })
    const renderless = wrapper.vm.$children[0]

    // uploading
    uploader.methods['getInProgress'] = jest.fn(() => 1)
    renderless.onStatusChange(0, uploader.qq.status.QUEUED, uploader.qq.status.UPLOADING)

    expect(wrapper).toMatchSnapshot()
    expect(wrapper.attributes().hidden).toBeUndefined()

    // done uploading
    uploader.methods['getInProgress'] = jest.fn(() => 0)
    renderless.onStatusChange(0, uploader.qq.status.UPLOADING, uploader.qq.status.UPLOAD_SUCCESSFUL)

    expect(wrapper).toMatchSnapshot()
    expect(wrapper.attributes().hidden).toBeUndefined()
  })

  it('shows file progress bar after all upload is complete if requested', () => {
    const uploader = new FineUploader({ options: {} })
    const wrapper = mount(ProgressBar, {
      propsData: { id: 0, uploader, hidesWhenNotUploading: false },
    })
    const renderless = wrapper.vm.$children[0]

    // uploading
    renderless.onStatusChange(0, uploader.qq.status.QUEUED, uploader.qq.status.UPLOADING)

    expect(wrapper).toMatchSnapshot()
    expect(wrapper.attributes().hidden).toBeUndefined()

    // done uploading
    renderless.onStatusChange(0, uploader.qq.status.UPLOADING, uploader.qq.status.UPLOAD_SUCCESSFUL)

    expect(wrapper).toMatchSnapshot()
    expect(wrapper.attributes().hidden).toBeUndefined()
  })
})
