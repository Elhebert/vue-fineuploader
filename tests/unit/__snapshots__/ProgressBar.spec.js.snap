// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProgressBar.vue hides file progress bar after all upload is complete 1`] = `<div aria-valuenow="0" aria-valuemax="100" aria-valuemin="0" role="progressbar" style="width: 0%;"></div>`;

exports[`ProgressBar.vue hides file progress bar after all upload is complete 2`] = `<div aria-valuenow="0" aria-valuemax="100" aria-valuemin="0" role="progressbar" style="width: 0%;" hidden="hidden"></div>`;

exports[`ProgressBar.vue hides the progress bar by default 1`] = `<div hidden="hidden" aria-valuenow="0" aria-valuemax="100" aria-valuemin="0" role="progressbar" style="width: 0%;"></div>`;

exports[`ProgressBar.vue hides the progress bar by default 2`] = `<div hidden="hidden" aria-valuenow="0" aria-valuemax="100" aria-valuemin="0" role="progressbar" style="width: 0%;"></div>`;

exports[`ProgressBar.vue hides total progress bar after all uploads are complete 1`] = `<div aria-valuenow="0" aria-valuemax="100" aria-valuemin="0" role="progressbar" style="width: 0%;"></div>`;

exports[`ProgressBar.vue hides total progress bar after all uploads are complete 2`] = `<div aria-valuenow="0" aria-valuemax="100" aria-valuemin="0" role="progressbar" style="width: 0%;" hidden="hidden"></div>`;

exports[`ProgressBar.vue show the file progress if an ID is supplied 1`] = `<div hidden="hidden" aria-valuenow="10" aria-valuemax="100" aria-valuemin="0" role="progressbar" style="width: 10%;"></div>`;

exports[`ProgressBar.vue show the total progress if no ID is supplied 1`] = `<div hidden="hidden" aria-valuenow="10" aria-valuemax="100" aria-valuemin="0" role="progressbar" style="width: 10%;"></div>`;

exports[`ProgressBar.vue shows file progress bar after all upload is complete if requested 1`] = `<div aria-valuenow="0" aria-valuemax="100" aria-valuemin="0" role="progressbar" style="width: 0%;"></div>`;

exports[`ProgressBar.vue shows file progress bar after all upload is complete if requested 2`] = `<div aria-valuenow="0" aria-valuemax="100" aria-valuemin="0" role="progressbar" style="width: 0%;"></div>`;

exports[`ProgressBar.vue shows the file progress bar before uploading start if specified 1`] = `<div aria-valuenow="0" aria-valuemax="100" aria-valuemin="0" role="progressbar" style="width: 0%;"></div>`;

exports[`ProgressBar.vue shows the total progress bar before uploading start if specified 1`] = `<div aria-valuenow="0" aria-valuemax="100" aria-valuemin="0" role="progressbar" style="width: 0%;"></div>`;

exports[`ProgressBar.vue shows total progress bar after all uploads are complete if requested 1`] = `<div aria-valuenow="0" aria-valuemax="100" aria-valuemin="0" role="progressbar" style="width: 0%;"></div>`;

exports[`ProgressBar.vue shows total progress bar after all uploads are complete if requested 2`] = `<div aria-valuenow="0" aria-valuemax="100" aria-valuemin="0" role="progressbar" style="width: 0%;"></div>`;
